test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + testJarPath);
+        // the content would have been extracted to current dir
+        verifyExtractedContent(".");
+    }
+
+    /**
+     * Tests that extracting a jar using {@code -P} flag and without any explicit destination
+     * directory works correctly if the jar contains entries with leading slashes and/or {@code ..}
+     * parts preserved.
+     */
+    @Test
+    public void testExtractNoDestDirWithPFlag() throws Exception {
+        // create a jar which has leading slash (/) and dot-dot (..) preserved in entry names
+        final Path jarPath = createJarWithPFlagSemantics();
+        // extract with -P flag without any explicit destination directory (expect the extraction to work fine)
+        final String[] args = new String[]{"-xvfP", jarPath.toString()};
+        printJarCommand(args);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, args);
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + jarPath);
+        final String dest = ".";
+        Assert.assertTrue(Files.isDirectory(Paths.get(dest)), dest + " is not a directory");
+        final Path d1 = Paths.get(dest, "d1");
+        Assert.assertTrue(Files.isDirectory(d1), d1 + " directory is missing or not a directory");
+        final Path d2 = Paths.get(dest, "d1", "d2");
+        Assert.assertTrue(Files.isDirectory(d2), d2 + " directory is missing or not a directory");
+        final Path f1 = Paths.get(LEADING_SLASH_PRESERVED_ENTRY);
+        Assert.assertTrue(Files.isRegularFile(f1), f1 + " is missing or not a file");
+        Assert.assertEquals(Files.readAllBytes(f1), FILE_CONTENT, "Unexpected content in file " + f1);
+        final Path f2 = Paths.get("d1/d2/../f2.txt");
+        Assert.assertTrue(Files.isRegularFile(f2), f2 + " is missing or not a file");
+        Assert.assertEquals(Files.readAllBytes(f2), FILE_CONTENT, "Unexpected content in file " + f2);
+    }
+
+    /**
+     * Tests that the {@code -P} option cannot be used during jar extraction when the {@code -C} and/or
+     * {@code --dir} option is used
+     */
+    @Test
+    public void testExtractWithDirPFlagNotAllowed() throws Exception {
+        final String expectedErrMsg = "-P option cannot be used when extracting a jar to a specific location";
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final List<String[]> cmdArgs = new ArrayList<>();
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", "."});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", "."});
+        cmdArgs.add(new String[]{"-xvfP", testJarPath.toString(), "-C", tmpDir});
```

Comment:
Perhaps add a DataProvider so you can test --extract as well?

Reply:
Not a data provider but the latest version of PR tests `--extract` as well.